content/blog_posts/articles/hmr-in-adonisjs.md

```diff
@@ -0,0 +1,106 @@
+
+This is a quick summary of how Hot Hook works, there is a bit more to it, but this is the main idea.
+
+Feel free to dive into the [Hot-Hook README](https://github.com/Julien-R44/hot-hook) for more information.
+
+## Using HMR in AdonisJS
+
+To use HMR in AdonisJS, you'll first need to install version `6.x.x` of `@adonisjs/core`.
+
+Then, also install `hot-hook`:
+
+```bash
+npm install -D hot-hook
+```
+
+Once done, you need to add the following configuration to your `package.json`
+
+```jsonc
+// title: package.json
+{
+  "name": "adonis-app",
+  "type": "module",
+  // ...
+  "hotHook": {
+    "boundaries": [
+      "./app/controllers/**/*.ts"
+    ]
+  }
+}
+```
+
+Then run `node ace serve --hmr` to start the dev server. And that's it! Try modifying a controller and you'll see that the server does not restart, and if you execute a request, the changes will be taken into account: you'll always have the latest version of your code.
+
+For more information on using it in AdonisJS, you can check out the [documentation page](https://docs.adonisjs.com/hot-module-reloading).
+
+## Why HMR?
+
+Some might ask, why add HMR to AdonisJS? Here are a few reasons why we decided to add this feature.
+
+### TSX as a template engine
+
+There's increasing talk about [using TSX as a template engine](https://adonisjs.com/blog/use-tsx-for-your-template-engine) in AdonisJS, as an alternative to Edge.
+
+With Edge, we don't need HMR, because Edge files are not JavaScript modules, they are not directly `import`ed and thus are not cached. Basically, to render an Edge view, we read the `.edge` file with `readFile` then compile it. Each time. So, we always have the latest version of the file when a view is rendered.
+
+However, with TSX: we're working with JavaScript modules. And, when a JavaScript module is imported, it gets cached. No matter how many times you modify the file, the imported module will always remain the same until the server is restarted.
+
+As a result, even a simple modification of padding/margin in our TSX view would cause a complete server restart, which can be long and tedious when developing your UI, for example.
+
+That's where HMR comes in: now, a modification of our TSX view just causes a page refresh, without server restart, and it's almost instantaneous. We now have an even better DX with TSX. **Really, it's a game changer.**
+
+### Large applications
+
+For applications that use a lot of packages (and thus `providers`), server restarts can be slows. The more providers you add, the more code there is to import and initialize at application startup. And the boot time can quickly reach several seconds: again, for the most impatient among us, this can be a bit frustrating development experience.
```

Comment:
```suggestion
Server restarts can be slow for applications that use many packages. The more providers you add, the more code there is to import and initialize at application startup. And the boot time can quickly reach several seconds: this can be a bit frustrating development experience.
```